### DISCUSSION TITLE
pybricksdev flash robot inventor hub over bluetooth

### DISCUSSION THREAD

**Discussion by mrinalraghupathi** [pybricks/support#1407] [Q&A]
Hello, 

Is it possible to flash the firmware on the robot inventor hub over bluetooth or does the hub need to be connected by USB? If it is connected by USB, does it need to be in DFU mode?

Thanks

**Comment by laurensvalk** [comment]
We could theoretically implement something like that, and I agree that it would be more convenient.

But doing so means that a successful update requires some firmware to be already on the hub, which is a bit of a catch 22.

DFU is the only builtin updater that is always on the hub, no matter what. So while it is a bit inconvenient, using DFU means that if users are able to install Pybricks, they are also always able to restore the original, because it works the same way.

So this way we can be 100% certain that nobody is ever left stuck with Pybricks :smile: 

And this way it also works for people coming from the MINDSTORMS app or hub.

It would still be nice to make it more convenient, as long as we can find a way to make it 100% fool proof.

**Reply by mrinalraghupathi** [reply]
Thanks for the explanation. 

Just to make sure I follow flashing over BLE does work for some hubs. Looking at this [code](https://github.com/pybricks/pybricksdev/blob/11667cb05427b2638fb475c1561fdfa380f59998/pybricksdev/cli/flash.py#L471) my understanding is that Boost, City and the Technic 4-port hub can be flashed over Bluetooth?

Just to check is the best way for spike prime/RI is what is mentioned [here](https://github.com/pybricks/pybricksdev/blob/master/README_dfu.rst) (DFU mode, backup, flash)

**Reply by laurensvalk** [reply]
Ah, yes, that's right. It isn't so much about the connectivity mode per se, more about whatever happens to be builtin into the bootloader made by LEGO.

For technic/city/boost (which don't have usb) this is BLE. For Spike/mindstorms this is DFU over USB.
